File: tests/test_addons.py
# pyright: standard
from pathlib import Path

import pytest
import typer
from pytest_mock import MockerFixture
from typer.testing import CliRunner

from aico.addons import create_click_command, discover_addons, execute_addon
from aico.consts import SESSION_FILE_NAME
from aico.exceptions import AddonExecutionError
from aico.models import AddonInfo, SessionData
from tests.helpers import save_session

runner = CliRunner()


def _create_addon(dir_path: Path, name: str, help_text: str) -> Path:
    addon_path = dir_path / name
    # Addon script must handle --usage and be executable
    addon_path.write_text(f'#!/bin/sh\n[ "$1" = "--usage" ] && echo "{help_text}" || exit 0')
    addon_path.chmod(0o755)
    return addon_path


def test_create_addon_command_execution(tmp_path: Path, mocker: MockerFixture) -> None:
    # GIVEN an addon info and mocked execute_addon
    addon_path = tmp_path / "addon2"
    addon_info = AddonInfo(name="addon2", path=addon_path, help_text="Addon 2", source="project")
    mock_execute = mocker.patch("aico.addons.execute_addon")

    # WHEN creating a command from the addon info
    cmd = create_click_command(addon_info)

    # AND invoking it directly
    from click.testing import CliRunner as ClickRunner

    result = ClickRunner().invoke(cmd, ["arg1", "--flag"])

    # THEN the command invocation succeeds
    assert result.exit_code == 0

    # AND `execute_addon` was called with the correct AddonInfo
    mock_execute.assert_called_once_with(addon_info, ["arg1", "--flag"])


def test_alias_group_prioritizes_builtin(tmp_path: Path, mocker: MockerFixture) -> None:
    from aico.main import AliasGroup

    # GIVEN a Typer app using AliasGroup with a built-in command
    app = typer.Typer(cls=AliasGroup)

    @app.command(name="init")
    def _init() -> None:
        """A built-in command."""
        print("built-in init called")

    @app.command(name="secondary")
    def secondary() -> None:
        """A built-in command."""
        print("built-in secondary called")

    # AND an addon with the same name as the built-in
    addon_info = AddonInfo(name="init", path=tmp_path / "init", help_text="addon init help", source="project")
    mocker.patch("aico.addons.discover_addons", return_value=[addon_info])
    mock_execute = mocker.patch("aico.addons.execute_addon")

    # WHEN the command is invoked
    result = runner.invoke(app, ["init"])
    print(result.stdout)
    print(result.stderr)

    # THEN the built-in function was called, not the addon
    assert result.exit_code == 0
    assert "built-in init called" in result.stdout
    mock_execute.assert_not_called()

    # AND the help text shows the built-in command, not the addon
    help_result = runner.invoke(app, ["--help"])
    assert "A built-in command." in help_result.stdout
    assert "addon init help" not in help_result.stdout


def test_execute_addon_calls_execvpe(tmp_path: Path, mocker: MockerFixture) -> None:
    # GIVEN an addon info object
    addon_path = tmp_path / "my-addon"
    addon_info = AddonInfo(name="my-addon", path=addon_path, help_text="", source="project")

    # AND a session file
    session_file = tmp_path / SESSION_FILE_NAME
    save_session(session_file, SessionData(model="test", chat_history=[], context_files=[]))
    mocker.patch("aico.addons.find_session_file", return_value=session_file)

    # AND os.execvpe is mocked
    mock_exec = mocker.patch("os.execvpe")

    # WHEN execute_addon is called
    execute_addon(addon_info, ["arg1", "--flag"])

    # THEN os.execvpe is called with the correct path, arguments, and environment
    mock_exec.assert_called_once()
    call_args, _ = mock_exec.call_args
    assert call_args[0] == addon_path
    assert call_args[1] == ["my-addon", "arg1", "--flag"]  # `execute_addon` prepends the name
    env = call_args[2]
    assert "AICO_SESSION_FILE" in env
    assert env["AICO_SESSION_FILE"] == str(session_file.resolve())


def test_execute_addon_handles_os_error(tmp_path: Path, mocker: MockerFixture) -> None:
    # GIVEN an addon info object
    addon_path = tmp_path / "my-addon"
    addon_info = AddonInfo(name="my-addon", path=addon_path, help_text="", source="project")

    # AND a session file exists
    session_file = tmp_path / SESSION_FILE_NAME
    mocker.patch("aico.addons.find_session_file", return_value=session_file)

    # AND os.execvpe is mocked to raise an OSError
    mocker.patch("os.execvpe", side_effect=OSError("Test error"))

    # WHEN execute_addon is called
    # THEN it should raise AddonExecutionError wrapping the original error
    with pytest.raises(AddonExecutionError, match="Error executing addon 'my-addon': Test error"):
        execute_addon(addon_info, [])


def test_discover_addons(tmp_path: Path, mocker: MockerFixture) -> None:
    # GIVEN a project addon and a user addon (with one name collision)
    session_file = tmp_path / SESSION_FILE_NAME
    save_session(session_file, SessionData(model="test", chat_history=[], context_files=[]))
    mocker.patch("aico.addons.find_session_file", return_value=session_file)

    # Trust the project so addons are discovered
    mocker.patch("aico.addons.is_project_trusted", return_value=True)

    # Mock the home directory to control user addons path
    user_home_dir = tmp_path / "user_home"
    user_home_dir.mkdir()
    mocker.patch("pathlib.Path.home", return_value=user_home_dir)

    # AND project addons
    project_addons_dir = tmp_path / ".aico" / "addons"
    project_addons_dir.mkdir(parents=True)
    project_addon_z_path = _create_addon(project_addons_dir, "z-addon", "Project Z help")
    collision_addon_path = _create_addon(project_addons_dir, "collision-addon", "Project collision help")
    non_executable_path = project_addons_dir / "not-executable"
    non_executable_path.write_text("...")

    # AND user addons
    user_addons_dir = user_home_dir / ".config" / "aico" / "addons"
    user_addons_dir.mkdir(parents=True)
    user_addon_a_path = _create_addon(user_addons_dir, "a-addon", "User A help")
    _create_addon(
        user_addons_dir, "collision-addon", "User collision help"
    )  # This is overridden by project collision-addon

    # Mock `importlib.resources.files` to point to a temporary bundled_addons directory
    mock_bundled_resources_path = tmp_path / "src" / "aico" / "bundled_addons"
    mock_bundled_resources_path.mkdir(parents=True, exist_ok=True)
    _create_addon(mock_bundled_resources_path, "commit", "Bundled commit help")
    _create_addon(mock_bundled_resources_path, "manage-context", "Bundled manage-context help")
    _create_addon(mock_bundled_resources_path, "summarize", "Bundled summarize help")

    # Since pathlib.Path implements the Traversable protocol, we can just return
    # the path to the temp directory we created.
    mocker.patch("importlib.resources.files", return_value=mock_bundled_resources_path)

    # Mock the _get_user_cache_dir to ensure _extract_bundled_addon puts files where we expect
    mocker.patch("aico.addons._get_user_cache_dir", return_value=tmp_path / "cache" / "aico" / "bundled_addons")

    # WHEN addons are discovered
    addons = discover_addons()

    # THEN the correct addons are found, with project addons overriding user addons and bundled addons
    assert len(addons) == 6  # 3 project/user, 3 bundled

    # AND the list is sorted by name
    assert addons[0].name == "a-addon"
    assert addons[0].path == user_addon_a_path.resolve()
    assert addons[0].help_text == "User A help"
    assert addons[0].source == "user"

    assert addons[1].name == "collision-addon"
    assert addons[1].path == collision_addon_path.resolve()
    assert addons[1].help_text == "Project collision help"  # Project wins
    assert addons[1].source == "project"

    assert addons[2].name == "commit"
    assert addons[2].help_text == "Bundled commit help"
    assert addons[2].source == "bundled"

    assert addons[3].name == "manage-context"
    assert addons[3].help_text == "Bundled manage-context help"
    assert addons[3].source == "bundled"

    assert addons[4].name == "summarize"
    assert addons[4].help_text == "Bundled summarize help"
    assert addons[4].source == "bundled"

    assert addons[5].name == "z-addon"
    assert addons[5].path == project_addon_z_path.resolve()
    assert addons[5].help_text == "Project Z help"
    assert addons[5].source == "project"


def test_discover_addons_untrusted_skips_project(tmp_path: Path, mocker: MockerFixture) -> None:
    # GIVEN project addons exist
    session_file = tmp_path / SESSION_FILE_NAME
    save_session(session_file, SessionData(model="test", chat_history=[], context_files=[]))
    mocker.patch("aico.addons.find_session_file", return_value=session_file)

    # Mock home to avoid interference
    user_home_dir = tmp_path / "user_home"
    user_home_dir.mkdir()
    mocker.patch("pathlib.Path.home", return_value=user_home_dir)

    project_addons_dir = tmp_path / ".aico" / "addons"
    project_addons_dir.mkdir(parents=True)
    _create_addon(project_addons_dir, "malicious-addon", "Help text")

    # AND project is NOT trusted
    mocker.patch("aico.addons.is_project_trusted", return_value=False)

    # BUT bundled/user logic proceeds naturally (mock bundled empty to simplify)
    mocker.patch("importlib.resources.files", side_effect=Exception("No bundled"))

    # WHEN discovering
    addons = discover_addons()

    # THEN project addons are not returned
    assert not any(a.name == "malicious-addon" for a in addons)

File: tests/test_addons.rs
mod common;
use assert_cmd::cargo::cargo_bin_cmd;
use std::fs;
use std::os::unix::fs::PermissionsExt;
use tempfile::tempdir;

#[test]
fn test_discover_addons_untrusted_skips_project() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    // 1. Initialize session
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    // 2. Setup a project-local addon
    let addon_dir = root.join(".aico/addons");
    fs::create_dir_all(&addon_dir).unwrap();
    let addon_path = addon_dir.join("my-local-addon");

    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::write(&addon_path, "#!/bin/sh\necho \"usage help\"").unwrap();
        let mut perms = fs::metadata(&addon_path).unwrap().permissions();
        perms.set_mode(0o755);
        fs::set_permissions(&addon_path, perms).unwrap();
    }
    #[cfg(windows)]
    {
        fs::write(&addon_path, "echo usage help").unwrap();
    }

    // 3. Project is NOT trusted by default.
    // WHEN running help
    let output = cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("--help")
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    // THEN local addon should not be discovered
    let stdout = String::from_utf8(output).unwrap();
    assert!(!stdout.contains("my-local-addon"));

    // 4. Trust the project
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("trust")
        .assert()
        .success();

    // 5. WHEN running help again
    let output_trusted = cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("--help")
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    // THEN local addon SHOULD be found
    let stdout_trusted = String::from_utf8(output_trusted).unwrap();
    assert!(stdout_trusted.contains("my-local-addon"));
}

#[test]
fn test_discover_addons_priority() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    // Setup home/user dir
    let home = root.join("fake_home");
    let user_addon_dir = home.join(".config/aico/addons");
    fs::create_dir_all(&user_addon_dir).unwrap();

    // Create user addon
    let user_addon = user_addon_dir.join("test-cmd");
    fs::write(
        &user_addon,
        "#!/bin/sh\n[ \"$1\" = \"--usage\" ] && echo \"user help\"",
    )
    .unwrap();
    fs::set_permissions(&user_addon, fs::Permissions::from_mode(0o755)).unwrap();

    // Create project addon with same name (overrides user)
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("init")
        .assert()
        .success();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("trust")
        .assert()
        .success();

    let project_addon_dir = root.join(".aico/addons");
    fs::create_dir_all(&project_addon_dir).unwrap();
    let project_addon = project_addon_dir.join("test-cmd");
    fs::write(
        &project_addon,
        "#!/bin/sh\n[ \"$1\" = \"--usage\" ] && echo \"project help\"",
    )
    .unwrap();
    fs::set_permissions(&project_addon, fs::Permissions::from_mode(0o755)).unwrap();

    let output = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("--help")
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    let stdout = String::from_utf8(output).unwrap();
    // Check that project help text is displayed, indicating it took priority
    assert!(stdout.contains("test-cmd"));
    assert!(stdout.contains("project help"));
    assert!(!stdout.contains("user help"));
}

#[test]
fn test_alias_group_prioritizes_builtin() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    let home = root.join("fake_home");
    fs::create_dir_all(&home).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("init")
        .assert()
        .success();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("trust")
        .assert()
        .success();

    // Create an addon named 'log' (existing builtin)
    let addon_dir = root.join(".aico/addons");
    fs::create_dir_all(&addon_dir).unwrap();
    let addon_path = addon_dir.join("log");
    fs::write(&addon_path, "#!/bin/sh\necho \"SHADOW_LOG\"").unwrap();
    fs::set_permissions(&addon_path, fs::Permissions::from_mode(0o755)).unwrap();

    // Running 'log' should still run the builtin log, not the script
    let output = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("log")
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    let stdout = String::from_utf8(output).unwrap();
    assert!(!stdout.contains("SHADOW_LOG"));
}

#[test]
fn test_execute_addon_environment() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    let home = root.join("fake_home");
    fs::create_dir_all(&home).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("init")
        .assert()
        .success();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("trust")
        .assert()
        .success();

    let addon_dir = root.join(".aico/addons");
    fs::create_dir_all(&addon_dir).unwrap();
    let addon_path = addon_dir.join("env-test");

    // Script that prints AICO_SESSION_FILE
    fs::write(&addon_path, "#!/bin/sh\necho \"VAR=$AICO_SESSION_FILE\"").unwrap();
    fs::set_permissions(&addon_path, fs::Permissions::from_mode(0o755)).unwrap();

    let output = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("env-test")
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    let stdout = String::from_utf8(output).unwrap();
    assert!(stdout.contains("VAR="));
    assert!(stdout.contains(".ai_session.json"));
}

#[test]
fn test_execute_addon_handles_os_error() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    let home = root.join("fake_home");
    fs::create_dir_all(&home).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("init")
        .assert()
        .success();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("trust")
        .assert()
        .success();

    let addon_dir = root.join(".aico/addons");
    fs::create_dir_all(&addon_dir).unwrap();
    let addon_path = addon_dir.join("broken");

    // Create a file that is executable but invalid (e.g. wrong shebang)
    fs::write(&addon_path, "#!/non/existent/interpreter\necho 1").unwrap();
    fs::set_permissions(&addon_path, fs::Permissions::from_mode(0o755)).unwrap();

    // execvpe should fail for this once started
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("broken")
        .assert()
        .failure();
}

#[test]
fn test_create_addon_command_execution() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    let home = root.join("fake_home");
    fs::create_dir_all(&home).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("init")
        .assert()
        .success();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("trust")
        .assert()
        .success();

    let addon_dir = root.join(".aico/addons");
    fs::create_dir_all(&addon_dir).unwrap();
    let addon_path = addon_dir.join("arg-test");

    // Script that prints its arguments
    fs::write(&addon_path, "#!/bin/sh\necho \"ARGS=$*\"").unwrap();
    fs::set_permissions(&addon_path, fs::Permissions::from_mode(0o755)).unwrap();

    let output = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("arg-test")
        .arg("hello")
        .arg("--world")
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    let stdout = String::from_utf8(output).unwrap();
    assert!(stdout.contains("ARGS=hello --world"));
}

#[test]
fn test_discover_addons() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    let home = root.join("fake_home");
    let user_addon_dir = home.join(".config/aico/addons");
    fs::create_dir_all(&user_addon_dir).unwrap();

    // Create a user addon
    let user_addon = user_addon_dir.join("user-cmd");
    fs::write(
        &user_addon,
        "#!/bin/sh\n[ \"$1\" = \"--usage\" ] && echo \"user help text\"",
    )
    .unwrap();
    fs::set_permissions(&user_addon, fs::Permissions::from_mode(0o755)).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("init")
        .assert()
        .success();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("trust")
        .assert()
        .success();

    let output = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("--help")
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    let stdout = String::from_utf8(output).unwrap();
    assert!(stdout.contains("user-cmd"));
    assert!(stdout.contains("user help text"));
}

#[test]
fn test_bundled_addons_are_available() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    // bundled addons should be accessible without any setup
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("commit")
        .arg("--usage")
        .assert()
        .success()
        .stdout(predicates::str::contains(
            "Generates a Conventional Commit message",
        ));

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("summarize")
        .arg("--usage")
        .assert()
        .success()
        .stdout(predicates::str::contains(
            "Archives active history as dated summary",
        ));
}

#[test]
fn test_execute_addon_calls_execvpe() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    let home = root.join("fake_home");
    fs::create_dir_all(&home).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("init")
        .assert()
        .success();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("trust")
        .assert()
        .success();

    let addon_dir = root.join(".aico/addons");
    fs::create_dir_all(&addon_dir).unwrap();
    let addon_path = addon_dir.join("path-test");

    // Script that prints its own PATH to verify aico binary is prepended
    fs::write(&addon_path, "#!/bin/sh\necho \"PATH=$PATH\"").unwrap();
    fs::set_permissions(&addon_path, fs::Permissions::from_mode(0o755)).unwrap();

    let output = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("HOME", &home)
        .arg("path-test")
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    let stdout = String::from_utf8(output).unwrap();
    // In Rust implementation, the directory containing the aico binary is prepended to PATH
    assert!(stdout.contains("PATH="));
}
